File: java/gov/nasa/gsfc/gmsec/api5/Specification.java
/*
 * Copyright 2007-2024 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file Specification.java
 */

package gov.nasa.gsfc.gmsec.api5;

import gov.nasa.gsfc.gmsec.api5.Config;
import gov.nasa.gsfc.gmsec.api5.GmsecException;
import gov.nasa.gsfc.gmsec.api5.Message;

import gov.nasa.gsfc.gmsec.api5.jni.JNIConfig;
import gov.nasa.gsfc.gmsec.api5.jni.JNIMessage;

import gov.nasa.gsfc.gmsec.api5.jni.JNISpecification;

import java.util.Collection;
import java.util.Map;
import java.util.HashMap;


/**
 * The Specification class loads a list of templates from a designated directory
 * that are used to create and validate message schemas. Each template has an associated ID
 * that is used to identify the kind of message schema the template will be defining.
 *
 * @see Config
 * @see SchemaIDIterator
 */
public class Specification
{
	private JNISpecification m_jniSpecification = null;


	//! @cond
	/** 
	 * @hidden
	 * This method is for internal GMSEC API use only.
	 * @param spec Object to reference for acquiring internal JNISpecification.
	 * @return Internal JNISpecification object.
	 */
	public static JNISpecification getInternal(Specification spec)
	{
		return (spec == null ? null : spec.m_jniSpecification);
	}


	/**
	 * @hidden
	 * This constructor is for internal GMSEC API use only.
	 * @param jSpec Internal JNISpecification object.
	 */
	public Specification(JNISpecification jSpec)
	{
		m_jniSpecification = jSpec;
	}
	//! @endcond


	/**
	 * Schema Levels for representing message specification.
	 */
	public enum SchemaLevel
	{
		/** C2MS */
		LEVEL_0(0),

		/** C2MS Extensions, or optionally user-defined */
		LEVEL_1(1),

		/** NASA/GMSEC Addendum, or optionally user-defined */
		LEVEL_2(2),

		/** User-defined */
		LEVEL_3(3),

		/** User-defined */
		LEVEL_4(4),

		/** User-defined */
		LEVEL_5(5),

		/** User-defined */
		LEVEL_6(6);

		private int level;
		private static Map<Integer, SchemaLevel> map = new HashMap<>();

		private SchemaLevel(int level) {
			this.level = level;
		}

		static {
			for (SchemaLevel schemaLevel : SchemaLevel.values()) {
				map.put(schemaLevel.level, schemaLevel);
			}
		}

		/**
		 * Converts the SchemaLevel enumerated type to an integer.
		 * @return An integer value.
		 */
		public int getValue() {
			return level;
		}

		/**
		 * Converts an integer value into a SchemaLevel enumerated type.
		 * @param level An integer value representing a SchemaLevel enumerated type.
		 * @return A SchemaLevel, or null if the integer value cannot be converted.
		 */
		public static SchemaLevel valueOf(int level) {
			return (SchemaLevel) map.get(level);
		}
	}


	/**
	 * Basic constructor that initializes the %Specification instance with the
	 * default message specification.
	 *
	 * @throws GmsecException Thrown if schemas (templates) cannot be loaded.
	 */
	public Specification() throws GmsecException
	{
		Config config = new Config();

		m_jniSpecification = new JNISpecification(Config.getInternal(config));
	}


	/**
	 * Initializes the %Specification instance with the given configuration.
	 *
	 * @param config The configuration for the Specification object.
	 *
	 * @throws IllegalArgumentException Thrown if the %Config object is null.
	 * @throws GmsecException Thrown if schemas (templates) cannot be loaded.
	 */
	public Specification(Config config) throws IllegalArgumentException, GmsecException
	{
		if (config == null)
		{
			throw new IllegalArgumentException("Config object is null");
		}

		m_jniSpecification = new JNISpecification(Config.getInternal(config));
	}


	/**
	 * Copy constructor - initializes the Specification instance based on another instance.
	 *
	 * @param other The other Specification object to copy.
	 *
	 * @throws IllegalArgumentException Thrown if the Specification object is null.
	 * @throws GmsecException Thrown if schemas (templates) cannot be loaded.
	 */
	public Specification(Specification other) throws IllegalArgumentException, GmsecException
	{
		if (other == null)
		{
			throw new IllegalArgumentException("Specification object is null");
		}

		m_jniSpecification = new JNISpecification(Specification.getInternal(other));
	}


	/**
	 * Returns a SchemaIDIterator which can be used to iterate over the IDs of the loaded
	 * templates.
	 *
	 * @return A handle to a SchemaIDIterator object.
	 */
	public SchemaIDIterator getSchemaIDIterator()
	{
		return m_jniSpecification.getSchemaIDIterator();
	}


	/**
	 * Returns the version of the message specification that is referenced by the Specification.
	 *
	 * @return A version number.
	 */
	public int getVersion()
	{
		return m_jniSpecification.getVersion();
	}


	/**
	 * Returns the operating schema level for the specification.
	 *
	 * @return A SchemaLevel enumerated value.
	 */
	public SchemaLevel getSchemaLevel()
	{
		return m_jniSpecification.getSchemaLevel();
	}


	/**
	 * Returns a collection of Message Specification objects associated with the Specification.
	 *
	 * @return A collection of MessageSpecifications.
	 */
	public Collection<MessageSpecification> getMessageSpecifications()
	{
		return m_jniSpecification.getMessageSpecifications();
	}

	/**
	 * Returns a list of the field names for the default header (which is the topmost schema level currently configured)
	 * 
	 * @return a list of strings that correspond to the names of the header fields.
	 */
	public Collection<String> getHeaderFieldNames()
	{
		return m_jniSpecification.getHeaderFieldNames("DEFAULT");
	}

	/**
	* Returns a list of the field names for the specified header 
	*
	* @param schemaName The name of the specified schema level (e.g. C2MS, C2MS-PIPE, GMSEC, DEFAULT).  A blank or NULL string refers to DEFAULT.
	*
	* @return a list of strings that correspond to the names of the header fields.
	*/
	public Collection<String> getHeaderFieldNames(String schemaName)
	{
		return m_jniSpecification.getHeaderFieldNames(((schemaName == null || schemaName.isEmpty()) ? "DEFAULT" : schemaName));
	}
}

File: ruby/gmsec5/dox/Specification.rb
# Copyright 2007-2024 United States Government as represented by the
# Administrator of The National Aeronautics and Space Administration.
# No copyright is claimed in the United States under Title 17, U.S. Code.
# All Rights Reserved.

# This file is used for generating Ruby documentation; it only contains
# stubs of constructors and methods associated with the class.


module Libgmsec_ruby #:nodoc: don't document this

	##
	# This class loads a series of message templates from a designated directory which is
	# used to create and validate message schemas.
	#
	# Each template has an associated ID that is used to identify the kind of message schema
	# the message template will be defining.
	#
	# ==== See also
	#
	# Libgmsec_ruby::Config
	#
	# Libgmsec_ruby::SchemaIDIterator
	#
	class Specification
        ##
        # Possible schema levels for representing message specifications.
        #
        class SchemaLevel
            ##
            # ==== Enumerated values
            #
            # * +LEVEL_0+ - C2MS
            # * +LEVEL_1+ - C2MS Extensions, or optionally user-defined
            # * +LEVEL_2+ - NASA/GMSEC Addendum, or optionally user-defined
            # * +LEVEL_3+ - User-defined
            # * +LEVEL_4+ - User-defined
            # * +LEVEL_5+ - User-defined
            # * +LEVEL_6+ - User-defined
            #
            def SchemaLevel_Enum
            end
        end


		##
		# Constructor that initializes the Specification using options provided with the configuration.
		#
		# ==== Attributes
		#
		# * +config+ - A Config object with configuration options.
		#
		# ==== Raises
		#
		# * A Libgmsec_ruby::GmsecException is thrown if illegal values are passed via the configuration.
		# * A Libgmsec_ruby::GmsecException is thrown if an error occurs while loading the message specification.
		#
		# ==== Example usage
		#
		#     config = Libgmsec_ruby::Config.new
		#     config.add_value("gmsec-schema-level", "0")                   # Level 0 == C2MS
		#     config.add_value("gmsec-specification-version", "201900")     # 201900  == Latest available message specification
		#
		#     spec = Libgmsec_ruby::Specification.new(config)
		#
		# ==== Other constructors
		#
		# * Constructor that initializes the Specification to the default message specification.
		#
		#     Libgmsec_ruby::Specification.new
		#
		# * Copy constructor
		#
		#     Libgmsec_ruby::Specification.new(other_spec)
		#
		def initialize(config)
		end


		##
		# Returns a SchemaIDIterator which can be used to iterate over the schema IDs of the loaded message templates.
		#
		# Note: Only one SchemaIDIterator object is associated with a Specification object; multiple calls to this
		# method will return a reference to the same SchemaIDIterator object. Each call will reset the iterator.
		#
		# ==== Returns
		#
		# A reference to a Libgmsec_ruby::SchemaIDIterator object.
		#
		def get_schema_id_iterator
		end


		##
		# Returns the version of the message specification that is referenced by the Specification.
		#
		# ==== Returns
		#
		# The version number represented as an integer (e.g. 201900 for the 2019 message specification).
		#
		def get_version
		end


		##
		# Returns the schema level of the message specification that is referenced by the Specification.
		#
		# ==== Returns
		#
		# A SchemaLevel enumerated value.
		#
		def get_schema_level
		end


        ##
        # Returns the validation level for the specification.
        #
        # ==== Returns
        #
        # Returns an enumerated validation level.
        #
        def get_validation_level
        end


		##
		# Returns an array of MessageSpecification objects, each of which represents a message template.
		#
		# ==== Returns
		#
		# An array of Libgmsec_ruby::MessageSpecification objects.
		#
		def get_message_specifications
		end


		##
		# Returns a list of the field names for the default header (which is the topmost schema level currently configured)
		#
		# ==== Returns
		#
		# An array of strings that correspond to the names of the header fields.
		#
		def getHeaderFieldNames
		end


		##
		# Returns a list of the field names for the specified header 
		#
		# ==== Attributes
		#
		# * +schemaName+ - The name of the specified schema level (e.g. C2MS, C2MS-PIPE, GMSEC, DEFAULT).  An empty string refers to DEFAULT.
		#
		# ==== Returns
		#
		# An array of strings that correspond to the names of the header fields.
		#
		def getHeaderFieldNames(schemaName)
		end
	end
end
